pacoloco.go

```diff
@@ -78,10 +144,78 @@ var (
 	downloadingFilesMutex sync.Mutex
 )
 
+// force resources prefetching
+func prefetchRequest(url string) (err error) {
+	urlPath := url
+	matches := pathRegex.FindStringSubmatch(urlPath)
+	if len(matches) == 0 {
+		return fmt.Errorf("input url path '%v' does not match expected format", urlPath)
+	}
+	repoName := matches[1]
+	path := matches[2]
+	fileName := matches[3]
+	repo, ok := config.Repos[repoName]
+	if !ok {
+		return fmt.Errorf("cannot find repo %s in the config file", repoName)
+	}
+	// create cache directory if needed
+	cachePath := filepath.Join(config.CacheDir, "pkgs", repoName)
+	if _, err := os.Stat(cachePath); os.IsNotExist(err) {
+		err := os.MkdirAll(cachePath, os.ModePerm)
+		if err != nil {
+			return err
+		}
+	}
+	filePath := filepath.Join(cachePath, fileName)
+	// mandatory update when prefetching,
+
+	mutexKey := repoName + ":" + fileName
+	downloadingFilesMutex.Lock()
+	fileMutex, ok := downloadingFiles[mutexKey]
+	if !ok {
+		fileMutex = &sync.Mutex{}
+		downloadingFiles[mutexKey] = fileMutex
+	}
+	downloadingFilesMutex.Unlock()
+	fileMutex.Lock()
+	defer func() {
+		fileMutex.Unlock()
+		downloadingFilesMutex.Lock()
+		delete(downloadingFiles, mutexKey)
+		downloadingFilesMutex.Unlock()
+	}()
+	// refresh the data in case if the file has been download while we were waiting for the mutex
+	ifLater := time.Time{} // spoofed to avoid rewriting downloadFile
+	downloaded := false
+	if repo.Url != "" {
+		downloaded, err = downloadFile(repo.Url+path+"/"+fileName, filePath, ifLater)
+		if err == nil && config.Prefetch != nil && !strings.HasSuffix(fileName, ".sig") && !strings.HasSuffix(fileName, ".db") {
+			updateDBPrefetchedFile(repoName, fileName) // update info for prefetching
+		}
+	} else {
+		for _, url := range repo.Urls {
+			downloaded, err = downloadFile(url+path+"/"+fileName, filePath, ifLater)
+			if err == nil {
+				if config.Prefetch != nil && !strings.HasSuffix(fileName, ".sig") && !strings.HasSuffix(fileName, ".db") {
+					updateDBPrefetchedFile(repoName, fileName) // update info for prefetching
+				}
+				break
+			}
+		}
+	}
+	if err != nil {
+		return err
+	} else if !downloaded {
+		return fmt.Errorf("not downloaded")
+	} else {
+		return nil
+	}
+}
+
 func handleRequest(w http.ResponseWriter, req *http.Request) error {
 	urlPath := req.URL.Path
 	matches := pathRegex.FindStringSubmatch(urlPath)
-	if matches == nil {
+	if len(matches) == 0 {
```

Comment:
len(matches) should have been used cause if matches is nil, returns 0, but if matches exists and matches nothing, it won't return nil